syntax = "proto3";
package metisfl;

import "metisfl/proto/model.proto";
import "metisfl/proto/learner.proto";
import "metisfl/proto/service_common.proto";

service ControllerService {
  rpc GetHealthStatus (Empty) returns (Ack) {}

  rpc SetInitialModel (Model) returns (Ack) {}

  rpc JoinFederation (Learner) returns (LearnerId) {}

  rpc LeaveFederation (LearnerId) returns (Ack) {}

  rpc StartTraining (Empty) returns (Ack) {}

  rpc TrainDone (TrainDoneRequest) returns (Ack) {}

  rpc GetLogs (Empty) returns (Logs) {}
  
  rpc ShutDown (Empty) returns (Ack) {}
}

message Learner {
  string hostname = 1;
  uint32 port = 2;
  string root_certificate_bytes = 3;
}

message LearnerId {
  string id = 1;
}

message TrainDoneRequest {
  Task task = 1;
  Model model = 3;
  TrainResults results = 4;
}

message TrainResults {
  string metrics = 1; // string representation of a JSON object
  map<string, double> metadata = 2;
}

message Logs {
  optional int32 global_iteration = 1;
  repeated Task tasks = 2;

  // Indexed by task_id
  map<string, TrainResults> train_results = 4;
  map<string, EvaluationResults> evaluation_results = 5;
  map<string, ModelMetadata> model_metadata = 6;
}

message ModelMetadata {
  double selection_duration_ms = 1;
  double aggregation_duration_ms = 2;
  repeated double aggregation_block_size = 3;
  repeated double aggregation_block_memory_kb = 4;
  repeated double aggregation_block_duration_ms = 5;
}
